device/transport/mqtt/src/mqtt.ts

```diff
@@ -765,46 +810,63 @@ export class Mqtt extends EventEmitter implements Client.Transport {
-            msg.properties.add(k, v);
-            break;
-        }
-      }
+      this._extractPropertiesFromTopicPart(topicParts[4], msg);
     }
 
     /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_004: [If there is a listener for the message event, a message event shall be emitted for each message received.]*/
     this.emit('message', msg);
   }
 
+  private _extractPropertiesFromTopicPart(properties: string, msg: Message): void {
+    const keyValuePairs = properties.split('&');
+
+    for (let i = 0; i < keyValuePairs.length; i++) {
+      const keyValuePair = keyValuePairs[i].split('=');
+      const k = decodeURIComponent(keyValuePair[0]);
+      const v = decodeURIComponent(keyValuePair[1]);
+
+      switch (k) {
+        case '$.mid':
+          /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_008: [When a message is received, the receiver shall populate the generated `Message` object `messageId` with the value of the property `$.mid` serialized in the topic, if present.]*/
+          msg.messageId = v;
+          break;
+        case '$.to':
+          /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_009: [When a message is received, the receiver shall populate the generated `Message` object `to` with the value of the property `$.to` serialized in the topic, if present.]*/
+          msg.to = v;
+          break;
+        case '$.exp':
+          /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_010: [When a message is received, the receiver shall populate the generated `Message` object `expiryTimeUtc` with the value of the property `$.exp` serialized in the topic, if present.]*/
+          msg.expiryTimeUtc = v;
+          break;
+        case '$.cid':
+          /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_011: [When a message is received, the receiver shall populate the generated `Message` object `correlationId` with the value of the property `$.cid` serialized in the topic, if present.]*/
+          msg.correlationId = v;
+          break;
+        case '$.uid':
+          /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_012: [When a message is received, the receiver shall populate the generated `Message` object `userId` with the value of the property `$.uid` serialized in the topic, if present.]*/
+          msg.userId = v;
+          break;
+        default:
+          /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_007: [When a message is received, the receiver shall populate the generated `Message` object `properties` property with the user properties serialized in the topic.]*/
+          msg.properties.add(k, v);
+          break;
+      }
+    }
+  }
+
+  private _onInputMessage(topic: string, payload: any): void {
+    /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_005: [When a message event is emitted, the parameter shall be of type Message]*/
+    let msg = new Message(payload);
+
+    const topicParts = topic.split('/');
+    if (topicParts[6]) {
+      this._extractPropertiesFromTopicPart(topicParts[6], msg);
+    }
+    let inputName: string = topicParts[5];
```

Comment:
at that point are we sure that topicParts has 6 sections and that inputName isn't going to be undefined?

Reply:
Yes, we're sure, messageRegex ensures this for us (I've modified it to make sure of this)

---
In reply to: [170389709](https://github.com/Azure/azure-iot-sdk-node/pull/235#discussion_r170389709) [](ancestors = 170389709)